servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpPayloadDiscardWatchdogServiceFilter.java

```diff
@@ -0,0 +1,84 @@
+
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.context.api.ContextMap;
+import io.servicetalk.http.api.HttpApiConversions;
+import io.servicetalk.http.api.HttpExecutionStrategies;
+import io.servicetalk.http.api.HttpExecutionStrategy;
+import io.servicetalk.http.api.HttpServiceContext;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+
+/**
+ * Filter which tracks HTTP payloads sent by the service so it can be freed if discarded in the pipeline.
+ */
+final class HttpPayloadDiscardWatchdogServiceFilter implements StreamingHttpServiceFilterFactory {
+
+    /**
+     * Instance of {@link HttpPayloadDiscardWatchdogServiceFilter}.
+     */
+    static final StreamingHttpServiceFilterFactory INSTANCE = new HttpPayloadDiscardWatchdogServiceFilter();
+
+    @SuppressWarnings("rawtypes")
+    static final ContextMap.Key<Publisher> payloadPublisherKey = ContextMap.Key
+            .newKey("io.servicetalk.http.netty.payloadPublisher", Publisher.class);
+
+    static final ContextMap.Key<Boolean> payloadSubscribedKey = ContextMap.Key
+            .newKey("io.servicetalk.http.netty.payloadSubscribed", Boolean.class);
+
+    private HttpPayloadDiscardWatchdogServiceFilter() {
+        // Singleton
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(final StreamingHttpService service) {
+
+        return new StreamingHttpServiceFilter(service) {
+            @Override
+            public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                        final StreamingHttpRequest request,
+                                                        final StreamingHttpResponseFactory responseFactory) {
+                return delegate()
+                        .handle(ctx, request, responseFactory)
+                        .map(response -> {
+                            if (!HttpApiConversions.isPayloadEmpty(response)) {
+                                // If the payload is not empty, always write the buffer publisher into the
+                                // request context. When a downstream subscriber arrives, mark the payload
+                                // as subscribed explicitly (having a payload present and no subscription is
+                                // an indicator that it must be freed later on).
+                                request.context().put(payloadPublisherKey, response.payloadBody());
+                                return response.transformPayloadBody(bufferPublisher ->
+                                        bufferPublisher.beforeOnSubscribe(subscription ->
```

Comment:
`beforeOnSubscribe` is a little late. `onSubscribed` is invoked by the publisher in response to `Publisher.subscribe` invocation from the `Subscriber`. Instead, consider using `beforeSubscriber`. Do the logic insider `Supplier`, and returns a noop Subscriber.

Ideally, we should add `beforeSubscribe(Runnable)` operator. We already discussed this with Scott, I need it for some other use-case. If you'd like to work on a new operator, I can guide you. Otherwise, just leave a todo comment and we can improve this when the new operator is available.